### ISSUE TITLE
filterNotNull() doesn't chain unsubscribe()

### ISSUE THREAD

**Issue by kubode** [ReactiveX/RxKotlin#38]
``` kotlin
val subscription = Observable.timer(2, TimeUnit.SECONDS)
    .filterNotNull()
    .subscribe { fail() }
subscription.unsubscribe()
```

`unsubscribe()` should stop Observable chain.
But, above code reaches to `fail()`.

If replace `filterNotNull()` to `filter { it != null }`, above code not reaches to `fail()`.

[Subscriber(Subscriber)](http://reactivex.io/RxJava/javadoc/rx/Subscriber.html#Subscriber%28rx.Subscriber%29) may fix this issue.

**Comment by zsxwing** [comment]
`requireNoNulls`, `withIndex` have similar issue as well.

**Comment by dsvoronin** [comment]
i use this extensions func in my projects, works well, can PR

```
public fun <T : Any> Observable<T?>.filterNotNull(): Observable<T> = filter { it != null }.map { it as T }
```

**Comment by zsxwing** [comment]
Go ahead :) You can also fix `requireNoNulls` and `withIndex` as well. Just use the existing operators instead of `lift`.

**Comment by zsxwing** [comment]
Resolved by #39
